### PR TITLE
Add IoSlice::as_bytes

### LINKED ISSUE #124659: `IoSlice`: the `Deref` impl is too restrictive on the lifetime of the slice, an alternative is needed.
I tried code similar to the following to advance an `std::io::IoSlice` ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=58a2946b0fae959b204a4148b5baa592)):

```rust
let mut buf = std::io::IoSlice::<'static>::new(b"foo");
let bytes: &[u8] = &buf;
let new_buf = std::io::IoSlice::new(&bytes[1..]);
buf = new_buf;
```
This does not compile, yielding the following error:

```
error[E0506]: cannot assign to `buf` because it is borrowed
 --> src/lib.rs:7:5
  |
5 |     let bytes: &[u8] = &buf;
  |                        ---- `buf` is borrowed here
6 |     let new_buf = std::io::IoSlice::new(&bytes[1..]);
7 |     buf = new_buf;
  |     ^^^^^^^^^^^^^
  |     |
  |     `buf` is assigned to here but it was already borrowed
  |     borrow later used here
```

The issue is that the `Defer<Target = [u8]>` impl on `IoSlice` discards the `'static` lifetime, replacing it with the lifetime of the of `buf`.

However, we should be able to write such code using only safe Rust: all we are doing is effectively storing a subslice of the original buffer. For example, if we replace `IoSlice` with normal slices `&[u8]` it compiles:
```rust
let mut buf: &'static [u8] = b"foo";
let bytes: &[u8] = &buf;
let new_buf = &bytes[1..];
buf = new_buf;
```

To make `IoSlice` more flexible there should be a method which returns a slice with the same lifetime of the underlying memory, something similar to the following:

```rust
impl<'a> IoSlice<'a> {
    pub fn as_bytes(&self) -> &'a [u8] {
        // ...
    }
}
```

> Note: this only applies to `IoSlice`, *not* `IoSliteMut` since it would break aliasing.

### Meta

`rustc --version --verbose`:
```
rustc 1.78.0 (9b00956e5 2024-04-29)
binary: rustc
commit-hash: 9b00956e56009bab2aa15d7bff10916599e3d6d6
commit-date: 2024-04-29
host: x86_64-unknown-linux-gnu
release: 1.78.0
LLVM version: 18.1.2
```

### PR DESCRIPTION
**Pull request by Lucretiel**
This PR proposes to add `IoSlice::as_bytes`. While `IoSlice` already has `Deref<Target=[u8]>`, this uses the lifetime of the `IoSlice` itself, throwing away the larger `'a` lifetime; `as_bytes` allows for getting the original slice with the original lifetime.

ACP: https://github.com/rust-lang/libs-team/issues/93
Fixes #124659

### CODE REVIEW

**Comment by Lucretiel** [comment]
My own personal use case for this is an attempt to write my own `write_all_vectored`. I ran into an issue where, given an `&mut [IoSlice<'a>]`, it's not possible to rewrite those slices in-place as the write proceeds, because slicing an `IoSlice` borrows it (instead of using the original `'a` lifetime), which precludes rewriting it in-place.

**Comment by Lucretiel** [comment]
Looks like the classic sorts of failures where an innate method is added and starts to take precedence over a trait method. Maybe `as_inner`, instead?

**Comment by joboet** [comment]
How about `into_bytes`? That fits well with methods like [`OccupiedEntry::into_mut`](https://doc.rust-lang.org/std/collections/hash_map/struct.OccupiedEntry.html#method.into_mut).

**Comment by Dylan-DPC** [comment]
@Lucretiel if you can address the CI failure and the conflicts, we can put this for review

**Comment by Lucretiel** [comment]
Taking a look now, thanks for your patience!

**Comment by Lucretiel** [comment]
Doctests are failing because of the use of an unstable library feature; can someone walk me through how to fix?

**Comment by aDotInTheVoid** [comment]
> can someone walk me through how to fix?

Add the `#feature` to the doctests. [Eg

**Comment by Mark-Simulacrum** [comment]
Can we add the same for IoSliceMut? It looks like it has the same lifetime issue.

**Comment by Lucretiel** [comment]
> Can we add the same for IoSliceMut? It looks like it has the same lifetime issue.

I'm actually pretty sure we can't, though I'd be happy to be proven wrong. A function resembling `fn as_bytes(&self) -> &'a [u8]` would end the `self` borrow when it returns, which would allow aliasing the mutable region. The best you can do is `into_bytes_mut(self) -> &'a mut [u8]`, which doesn't offer many advantages over just dropping the `IoSliceMut`.

**Comment by Lucretiel** [comment]
In any case, this is (finally) ready to merge, just needs approval from someone.

File: library/std/src/io/mod.rs [hunk 1]
```diff
@@ -1491,6 +1491,31 @@ impl<'a> IoSlice<'a> {
...
+    ///
+    /// let mut io_slice = IoSlice::new(data);
+    /// let tail = &io_slice.into_bytes()[3..];
+    ///
+    /// // This works because `tail` doesn't borrow `io_slice`
+    /// io_slice = IoSlice::new(tail);
+    ///
+    /// assert_eq!(io_slice.into_bytes(), b"def");
+    /// ```
+    #[unstable(feature = "io_slice_as_bytes", issue = "111277")]
```
Comment (tgross35) [inline library/std/src/io/mod.rs:1513]: The issue number points to this PR; you should open a tracking issue and link it instead.

Also feature gate `io_slice_as_bytes` -> `io_slice_into_bytes` if that is the name

**Comment by tgross35** [comment]
The ACP at https://github.com/rust-lang/libs-team/issues/93 covers the change here. Two changes from this PR are that the accepted name was `as_slice` (though presumably this is up for bikeshedding), and that it can be implemented for `IoSliceMut` too. Could you make these changes, as well as updating the `unstable` attribute as mentioned above?

@rustbot author

**Comment by Dylan-DPC** [comment]
@Lucretiel any updates on this? thanks

**Comment by alex-semenyuk** [comment]
@Lucretiel 
From wg-triage. Closed this PR due to inactivity. Feel free to reopen or raised new one. Thanks for your efforts.

**Pull request closed without merging**